cpp/daal/src/services/service_topo.cpp

```diff
@@ -54,219 +184,565 @@ namespace services
+    unsigned error;
+    // If CPUID full reporting capability has been restricted, we need to be aware of it.
+    unsigned Alert_BiosCPUIDmaxLimitSetting;
+
+    unsigned OSProcessorCount = 0; // how many logical processor the OS sees
+    bool hasLeafB;                 // flag to keep track of whether CPUID leaf 0BH is supported
+    unsigned maxCacheSubleaf;      // highest CPUID leaf 4 subleaf index in a processor
+
+    // the following global variables are the total counts in the system resulting from software enumeration
+    unsigned EnumeratedPkgCount;
+    unsigned EnumeratedCoreCount;
+    unsigned EnumeratedThreadCount;
+
+    // CPUID ID leaf 4 can report data for several cache levels, we'll keep track of each cache level
+    std::array<unsigned, MAX_CACHE_SUBLEAFS> EnumeratedEachCacheCount;
+    // the following global variables are parameters related to
+    //  extracting sub IDs from an APIC ID, common to all processors in the system
+    unsigned SMTSelectMask;
+    unsigned PkgSelectMask;
+    unsigned CoreSelectMask;
+    unsigned PkgSelectMaskShift;
+    unsigned SMTMaskWidth;
+    // We'll do sub ID extractions using parameters from each cache level
+    std::array<unsigned, MAX_CACHE_SUBLEAFS> EachCacheSelectMask;
+    std::array<unsigned, MAX_CACHE_SUBLEAFS> EachCacheMaskWidth;
+
+    // the following global variables are used for product capability identification
+    unsigned HWMT_SMTperCore;
+    unsigned HWMT_SMTperPkg;
+    // a data structure that can store simple leaves and complex subleaves of all supported leaf indices of CPUID
+    CPUIDinfox * cpuid_values = nullptr;
+    // workspace of our generic affinitymask structure to allow iteration over each logical processors in the system
+    GenericAffinityMask cpu_generic_processAffinity;
+    GenericAffinityMask cpu_generic_systemAffinity;
+    // workspeace to assist text display of cache topology information
+    std::array<cacheDetail_str, MAX_CACHE_SUBLEAFS> cacheDetail;
+
+    bool isInit = false;
+
+    glktsn();
+
+    ~glktsn() { freeArrays(); }
+
+private:
+    int allocArrays(const unsigned cpus);
+    void freeArrays();
+    unsigned getMaxCPUSupportedByOS();
+    int cpuTopologyParams();
+    int cpuTopologyLeafBConstants();
+    int cpuTopologyLegacyConstants(CPUIDinfo * pinfo, DWORD maxCPUID);
+    int cacheTopologyParams();
+    void initStructuredLeafBuffers();
+    int findEachCacheIndex(DWORD maxCPUID, unsigned cache_subleaf);
+    int eachCacheTopologyParams(unsigned targ_subleaf, DWORD maxCPUID);
+
+    void setChkProcessAffinityConsistency();
+    int initEnumeratedThreadCountAndParseAPICIDs();
+    int analyzeCPUHierarchy();
+    int analyzeEachCHierarchy(unsigned subleaf);
+    void buildSystemTopologyTables();
+};
+
+// Global CPU topology object
+static glktsn globalCPUTopology;
 
 static char scratch[BLOCKSIZE_4K]; // scratch space large enough for OS to write SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX
```

Comment:
```suggestion
static std::array<char,  BLOCKSIZE_4K> scratch; // scratch space large enough for OS to write SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX
```

For more debuggability.